### PR TITLE
Shibboleth

### PR DESCRIPTION
**Pull request by beckstefan**
This PR enables planned support for shibboleth.

To have the documentation better readable, it is setup on readthedocs while PR being not merged: https://dissemin.readthedocs.io/en/shibboleth/configuration.html#shibboleth

- [x] Set up shibboleth SP3
- [x] Patch  `shibboleth`-package to hook into creation of django users and researcher model
  - [x] Model to map django users with shibboleth eduPersonPrincipalName
  - [x] Create / update researcher profil after shibboleth login
- [x] Match shibboleth with orcid to orcid-user
- [x] Match orcid-user to shibboleth with orcid
- [x] Solve multiple emails (from orcid and shibboleth)
- [x] Make sure everything works if researcher does not have an orcid
- [x] Login template
  - [x] Documentation about discovery service
  - [x] Template + View
- [x] Add paragraph about data privacy
- [x] Logout
- [x] Determine repository based on shibboleth user data

### CODE REVIEW

**Comment by beckstefan** [comment]
With a new authentication method, we require of course a new login template. Below is a first draft. I am not completely happy and would appreciate feedback and further ideas.

Screenshot_2020-11-10 Sign in - Dissemin

**Comment by wetneb** [comment]
I find this very nice!

**Comment by beckstefan** [comment]
So another part is, of course, data privacy.

Below is my suggestestion:

# TOS

## In short

### Privacy

#### Old

The platform does not include any analytics, advertisement or tracking services. We are however using the <a href="https://sentry.io/privacy/">Sentry service</a> for error tracking on this platform. Our webserver configuration stores basic access information (URL requested and IP address) for error monitoring, which are kept for fourteen days.

Authenticated users are logged-in via ORCID, a third-party service. The platform has no access to the private information they provide to ORCID. In particular, we do not store any ORCID passwords, and we do not have access to email addresses that are marked as private in the ORCID record.

We need to set cookies to authenticate our users, therefore by logging in on this platform, users accept these cookies.

#### New

Authenticated user are logged-in either via eduGAIN/Shibboleth or ORCID.

Dissemin is a participant of <a href="https://www.renater.fr">RENATER</a> and participates in eduGAIN. We receive a persistant identifier and the name of the user. The email and orcid are optional. The identity providerwill give further information about the transmissioned data.

ORCID is a third-party service. The platform has no access to the private information they provide to ORCID. In particular, we do not store any ORCID passwords, and we do not have access to email addresses that are marked as private in the ORCID record.

## In full

### Personel Data

#### Old

Dissemin uses ORCID's public OAuth interface to authenticate its users. This interface only transmits information marked as public in the user's ORCID profile. This means that ORCID passwords are never transmitted and email addresses are not transmitted when they are marked as private, which is the default setting.
            
When logging in to Dissemin from ORCID, Dissemin acquires the user's ORCID id, as well as all public information attached to the ORCID record.

We are using the <a href="https://sentry.io/privacy/">Sentry service</a> for error tracking on this platform. Our webserver configuration stores basic access information (URL requested and IP address) for error monitoring, which are kept for fourteen days.

Dissemin can store any information input by its users on the platform. Users can ask for a copy of the personal information Dissemin stores about them by writing an email to <a href="mailto:support@dissem.in">support@dissem.in</a>.


#### New

Dissemin uses eduGAIN/Shibboleth and ORCID's public OAuth interface to authenticate its users. 

When authenticating with eduGAIN/Shibboleth, the user will authenticate against an identity provider participating in eduGAIN. Dissemin requires the following attributes to be released by the identity provider: eduPersonTargetedId (required), name (required), email (optional) and ORCID (optional). In case that a valid ORCID is transmitted, Dissemin will harvest public data from the corresponding ORCID profile

The ORCID interface only transmits information marked as public in the user's ORCID profile. This means that ORCID passwords are never transmitted and email addresses are not transmitted when they are marked as private, which is the default setting.
            
When logging in to Dissemin from ORCID, Dissemin acquires the user's ORCID id, as well as all public information attached to the ORCID record.

We are using the <a href="https://sentry.io/privacy/">Sentry service</a> for error tracking on this platform. Our webserver configuration stores basic access information (URL requested and IP address) for error monitoring, which are kept for fourteen days.

Dissemin can store any information input by its users on the platform. Users can ask for a copy of the personal information Dissemin stores about them by writing an email to <a href="mailto:support@dissem.in">support@dissem.in</a>.

Dissemin will transmit name, email or ORCID of the depositing user to the user-selected repository depending on the requirements of the selected repository.

# FAQ

I would add the following to the authentication section
 
## What is the difference between authentication via ORCID and eduGAIN?

If your identity provider sends Dissemin an ORCID, there is no difference in functionality, except that for some institutional repositories we can confirm that you are a member of the owning instution of the repository if you authenticate with eduGAIN---which may accelerate the publication process in the repository.

If your identity provide does not transmit an ORCID, we can't associate your account with an ORCID profile.

## I authenticate with eduGAIN and ORCID, can I combine the accounts into one?

Based on the ORCID, Dissemin will automatically merge both accounts into one. This requires that your identity provider releases your ORCID. If your identity provider supports this functionality, you can add your ORCID in the user management of your identity provider.

---

Any remarks and corrections are welcome!

**Comment by wetneb** [comment]
@beckstefan thanks for the heads up! Feel free to add your new TOS contents as committed changes in the PR, so that we can comment and suggest changes more easily than in a message.

**Comment by beckstefan** [comment]
The above version are now commited.

File: website/templates/dissemin/faq.html [hunk 1]
```diff
@@ -456,21 +460,32 @@ <h3 id="copyAvailable">{{ omission_available }}</h3>
             <h2>{{ auth }}</h2>
         </div>
         <div class="card-body">
-            <h3 id="orcidWhy">{{ auth }}</h3>
+            <h3 id="authWhy">{{ auth_why }}</h3>
             <p>
                 {% blocktrans trimmed %}
                     Authenticating users is necessary to deposit papers on their behalf in open repositories. Besides, knowing a user's ORCID id helps us find its publications, avoiding articles authored by namesakes for instance. Logging in via ORCID will not expose any private information to Dissemin: we only have access to the public part of your ORCID record. For instance, unless you have explicitly marked your email address as public, we do not even know your email address when you log in to Dissemin via ORCID.
+
+                    Authenticatoin with eduGAIN/Shibboleth exposes us only the information that your identity provider is willing to transfer.
```
Comment (wetneb) [inline website/templates/dissemin/faq.html:468]:
```suggestion
                    Authentication with eduGAIN/Shibboleth only exposes to us the information that your identity provider is willing to transfer.
```

File: website/templates/dissemin/tos.html [hunk 2]
```diff
@@ -28,14 +28,15 @@ <h3 id="scope">{% trans "Scope of the service" %}</h3>
             <h3 id="privacy">{% trans "Privacy" %}</h3>
             <p>
                 {% blocktrans trimmed %}
-                    The platform does not include any analytics, advertisement or tracking services. We are however using the <a href="https://sentry.io/privacy/">Sentry service</a> for error tracking on this platform. Our webserver configuration stores basic access information (URL requested and IP address) for error monitoring, which are kept for fourteen days.
+                    Authenticated user are logged-in either via eduGAIN/Shibboleth or ORCID.
```
Comment (wetneb) [inline website/templates/dissemin/tos.html:31]:
```suggestion
                    Authenticated users are logged-in either via eduGAIN/Shibboleth or ORCID.
```

File: website/templates/dissemin/tos.html [hunk 3]
```diff
@@ -28,14 +28,15 @@ <h3 id="scope">{% trans "Scope of the service" %}</h3>
...
+
+                    Dissemin is a participant of <a href="https://www.renater.fr">RENATER</a> and participates in eduGAIN. We receive a persistant identifier and the name of the user. The email and orcid are optional. The identity providerwill give further information about the transmissioned data.
```
Comment (wetneb) [inline website/templates/dissemin/tos.html:33]:
```suggestion
                    Dissemin is a participant of <a href="https://www.renater.fr">RENATER</a> and participates in eduGAIN. When a user logs in to Dissemin via eduGAIN, we receive a persistent identifier and the name of the user. The email address and ORCID identifier are optional. The identity provider can supply additional metadata which we do not rely on.
```

File: website/templates/dissemin/tos.html [hunk 4]
```diff
@@ -28,14 +28,15 @@ <h3 id="scope">{% trans "Scope of the service" %}</h3>
...
 
-Authenticated users are logged-in via ORCID, a third-party service.
-The platform has no access to the
-private information they provide to ORCID. In particular, we do
-not store any ORCID passwords, and we do not have access to email addresses that are marked as private in the ORCID record.
+                    ORCID is a third-party service. The platform has no access to the private information they provide to ORCID. In particular, we do not store any ORCID passwords, and we do not have access to email addresses that are marked as private in the ORCID record.
```
Comment (wetneb) [inline website/templates/dissemin/tos.html:35]:
```suggestion
                    Users can also log in via <a href="https://orcid.org">ORCID</a>, a third-party service. The platform has no access to the private information they provide to ORCID. In particular, we do not store any ORCID passwords, and we do not have access to email addresses that are marked as private in the ORCID record.
```

**Review by wetneb — APPROVED** [review]

**Pull request merged**